pkg/rbd/nodeserver.go

```diff
@@ -115,15 +115,25 @@ func (ns *NodeServer) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
 	}
 	klog.V(4).Infof("rbd image: %s/%s was successfully mapped at %s\n", req.GetVolumeId(), volOptions.Pool, devicePath)
 
+	// if mounting to stagingpath fails unmap the rbd device. this wont leave any
+	// stale rbd device if unstage is called
+	defer func() {
+		if err != nil {
+			// Unmapping rbd device
+			if err = detachRBDDevice(devicePath); err != nil {
+				klog.Errorf("failed to unmap rbd device: %s for volume %s with error: %v", devicePath, volID, err)
+			}
+		}
+	}()
 	// nodeStage Path
 	err = ns.mountVolumeToStagePath(req, stagingTargetPath, devicePath)
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 
-	err = os.Chmod(stagingTargetPath, 0777)
-	if err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
+	perErr := os.Chmod(stagingTargetPath, 0777)
```

Comment:
> If we return a different error (`perErr`) then the CO would assume NodeStage failed, and **may** retry.
> 
> Case retry: Retry would succeed at line 102, and hence would not have set the mode bits

lets unmap if chmod fails (chance of having this is very negligible)

> Case no-retry: Depending on what the CO does later, we would have this RBD image and mount leaked on the node

if CO won't try this won't leak a mount

> 
> The fix does improve the current situation, i.e unmap on mount failures, but needs some more attention in the future on other failures post this to avoid the above situations.

am adding some more validation in case of non-block let's wait.